java/gazelle/private/types/types.go

```diff
@@ -0,0 +1,85 @@
+package types
+
+import (
+	"fmt"
+	"strings"
+	"unicode"
+)
+
+type PackageName struct {
+	Name string
+}
+
+func NewPackageName(name string) PackageName {
+	return PackageName{
+		Name: name,
+	}
+}
+
+func PackageNameLess(l, r PackageName) bool {
+	return l.Name < r.Name
+}
+
+type ClassName struct {
+	packageName        PackageName
+	bareOuterClassName string
+	innerClassNames    []string
+}
+
+func (c *ClassName) PackageName() PackageName {
+	return c.packageName
+}
+
+func (c *ClassName) BareOuterClassName() string {
+	return c.bareOuterClassName
+}
+
+func (c *ClassName) FullyQualifiedClassName() string {
+	var parts []string
+	if c.packageName.Name != "" {
+		parts = append(parts, strings.Split(c.packageName.Name, ".")...)
+	}
+	parts = append(parts, c.bareOuterClassName)
+	parts = append(parts, c.innerClassNames...)
+	return strings.Join(parts, ".")
+}
+
+func NewClassName(packageName PackageName, bareOuterClassName string) ClassName {
+	return ClassName{
+		packageName:        packageName,
+		bareOuterClassName: bareOuterClassName,
+	}
+}
+
+func ParseClassName(fullyQualified string) (*ClassName, error) {
+	parts := strings.Split(fullyQualified, ".")
```

Comment:
I guess we've already handled the case where in Java an inner class is represented as something like `com.example.Foo$Bar`? We can rely on the fully qualified name just having period characters separating things?

Reply:
Yeah, we're only getting these values from source code itself, where $s aren't present; we don't currently parse e.g. filenames from inside jar files.